### ISSUE TITLE
Chess implementation?

### ISSUE THREAD

**Issue by jamesbraza** [suragnair/alpha-zero-general#296]
I am making this issue on 2/28/2023, and at this point in time I don't see a chess implementation.

Are there any plans to implement chess here?

Alternately, are there any known repo(s) that implement chess using this AGZ general framework?  Preferably they use the [`python-chess` library](https://github.com/niklasf/python-chess) to handle chess specifics.

---

**Research**

Based on issues in this repo, I have found multiple descendants:
- @TheMessik: [chess_game](https://github.com/TheMessik/alpha-zero-general/tree/master/chess_game), last commit on 12/17/2021
- @yeekit24: [yeekit24/alpha-zero-general](https://github.com/yeekit24/alpha-zero-general) has [two relevant branches](https://github.com/yeekit24/alpha-zero-general/compare/implement-chess...chess-implementation), last commit on 12/13/2021
- @JernejHenigman's [JernejHenigman/LosingChessAlphaZero](https://github.com/JernejHenigman/LosingChessAlphaZero), last commit on 11/7/2021
    - Note: [losing chess](https://en.wikipedia.org/wiki/Losing_chess) is a chess variant where you're trying to lose first
- @goshawk22: [goshawk22/alpha-zero-chess](https://github.com/goshawk22/alpha-zero-chess), last commit on 3/21/2021
    - Author has been active in many issues and this repo is mentioned [here](https://github.com/suragnair/alpha-zero-general/issues/241#issuecomment-917495205)

I also discovered this 4/18/2021 PR (currently unaccepted) by @yeekit24: https://github.com/suragnair/alpha-zero-general/pull/240

Lastly, here are a few other noteworthy repos, not based on this framework:
- [geochri/AlphaZero_Chess](https://github.com/geochri/AlphaZero_Chess)
- [Zeta36/chess-alpha-zero](https://github.com/Zeta36/chess-alpha-zero)

**Comment by TheMessik** [comment]
Mine is very much WIP and at an early stage. It should be runnable, but the network you get out of it isn't anywhere near good

**Comment by jamesbraza** [comment]
@TheMessik where did you get your network architecture from (e.g. use of batch norms, hyperparams of layers, etc)?

**Comment by TheMessik** [comment]
I believe I based it on the Othello network, adding and removing layers and changing parameters until it worked. That part was rushed and needs a rewrite

**Comment by vedanta28** [comment]
@TheMessik What do you think that it's Elo is? Like I wanted to create something for a variation of chess. Three Check Chess with an additional rule that pawns move diagonally and capture moving forward. Would just changing these rules in your chess.py and training it , work?

**Comment by TheMessik** [comment]
It's ELO will definitely be in single digits. I was training it on endgames, so it's opening and middlegame is nonexistent, it's pretty much a random player there.

I don't think python-chess supports 3 checks variant, so you'll also need to do your own game ended checking, aside from adapting the move masking function.

**Comment by jamesbraza** [comment]
Hello all, for posterity here is my working `chess` implementation: https://github.com/jamesbraza/cs234-dreamchess/tree/main/azg_chess

The AI inside [`nn.py`](https://github.com/jamesbraza/cs234-dreamchess/blob/main/azg_chess/nn.py) was trainable and shown to improve in skill across rounds of self-play. I found an "unsigned" embedding (read the module for what that means) worked better than a "signed" embedding given the same number of training iterations. Also, the [`Game`](https://github.com/jamesbraza/cs234-dreamchess/blob/main/azg_chess/game.py#L67) subclass is fully unit tested and works well.

For me, this was part of a [Stanford CS234 Reinforcement Learning](https://cs234.stanford.edu/) class project that I was rushing to complete in the last weeks, so I didn't have time to do a full-blown multi-day training to see how strong the AI could get (measured in Elo). Hope this can help someone in the future, or serve as a better starting point for a PR into this repo. Cheers!
